grug/app/src/execute.rs

```diff
@@ -566,24 +587,61 @@ where
     AppError: From<VM::Error>,
 {
     let chain_id = CHAIN_ID.load(&storage)?;
-    let mut contract_info = CONTRACTS.load(&storage, msg.contract)?;
+    let mut old_code_hash = None;
 
-    // Only the account's admin can migrate it
-    let Some(admin) = &contract_info.admin else {
-        return Err(AppError::AdminNotSet);
-    };
+    // Update the contract info.
+    let contract_info = CONTRACTS.update(&mut storage, msg.contract, |mut info| {
+        old_code_hash = Some(info.code_hash);
 
-    if sender != admin {
-        return Err(AppError::NotAdmin {
-            sender,
-            admin: contract_info.admin.unwrap(),
-        });
-    }
+        // Ensure the sender is the admin of the contract.
+        if Some(sender) != info.admin {
+            return Err(AppError::Unauthorized);
+        }
+
+        info.code_hash = msg.new_code_hash;
+
+        Ok(info)
+    })?;
+
+    // Reduce usage count of the old code.
+    CODES.update(
+        &mut storage,
+        old_code_hash.unwrap(),
+        |mut code| -> StdResult<_> {
```

Comment:
This is so ugly. I'm going to adjust rustfmt settings in a separate PR, so that these are put on the same line...